pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,184 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificaterequests
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+
+	"github.com/kr/pretty"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/validation"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	errorCertificatePending = "CertPending"
+	errorCertificateFailed  = "CertFailed"
+
+	errorIssuerNotFound = "IssuerNotFound"
+	errorIssuerInit     = "IssuerInitError"
+
+	successCertificateIssued = "CertIssued"
+)
+
+func (c *Controller) Sync(ctx context.Context, cr *v1alpha1.CertificateRequest) (err error) {
+	c.metrics.IncrementSyncCallCount(ControllerName)
+
+	log := logf.FromContext(ctx)
+	dbg := log.V(logf.DebugLevel)
+
+	crCopy := cr.DeepCopy()
+	defer func() {
+		if _, saveErr := c.updateCertificateRequestStatus(ctx, cr, crCopy); saveErr != nil {
+			err = utilerrors.NewAggregate([]error{saveErr, err})
+		}
+	}()
+
+	dbg.Info("fetching issuer object referenced by CertificateRequest")
+
+	issuerObj, err := c.helper.GetGenericIssuer(crCopy.Spec.IssuerRef, crCopy.Namespace)
+	if k8sErrors.IsNotFound(err) {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		log.WithValues(
+			logf.RelatedResourceNameKey, crCopy.Spec.IssuerRef.Name,
+			logf.RelatedResourceKindKey, crCopy.Spec.IssuerRef.Kind,
+		).Error(err, "failed to find referenced issuer")
+	}
+	if err != nil {
+		return err
+	}
+
+	dbg.Info("ensuring issuer type matches this controller")
+
+	log = logf.WithRelatedResource(log, issuerObj.GetObjectMeta())
+
+	issuerType, err := apiutil.NameForIssuer(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		log.Error(err, "failed to obtain referenced issuer type")
+		return nil
+	}
+
+	// This CertificateRequest is not meant for us, ignore
+	if issuerType != c.issuerType {
+		c.log.WithValues(
+			logf.RelatedResourceKindKey, issuerType,
+		).V(5).Info("issuer reference type does not match controller resource kind, ignoring")
+		return nil
+	}
+
+	dbg.Info("validating CertificateRequest resource object")
+
+	el := validation.ValidateCertificateRequest(crCopy)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	if len(crCopy.Status.Certificate) > 0 {
+		dbg.Info("certificate exists so exiting sync")
```

Comment:
nit: change to `certificate field is already set in status so skipping processing`?